src/util/threadpool.rs

```diff
@@ -0,0 +1,618 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::usize;
+use std::sync::{Arc, Mutex};
+use std::sync::mpsc::{channel, Sender, Receiver};
+use std::thread::{Builder, JoinHandle};
+use std::boxed::FnBox;
+use std::collections::{BinaryHeap, HashMap, VecDeque};
+use std::cmp::Ordering;
+use std::hash::Hash;
+use std::marker::PhantomData;
+
+pub struct Task<T> {
+    // The task's number in the pool.Each task has a unique number,
+    // and it's always bigger than preceding ones.
+    id: u64,
+    // the task's group_id.
+    group_id: T,
+    task: Box<FnBox() + Send>,
+}
+
+impl<T> Task<T> {
+    fn new<F>(id: u64, group_id: T, job: F) -> Task<T>
+        where F: FnOnce() + Send + 'static
+    {
+        Task {
+            id: id,
+            group_id: group_id,
+            task: Box::new(job),
+        }
+    }
+}
+
+impl<T> Ord for Task<T> {
+    fn cmp(&self, right: &Task<T>) -> Ordering {
+        self.id.cmp(&right.id).reverse()
+    }
+}
+
+impl<T> PartialEq for Task<T> {
+    fn eq(&self, right: &Task<T>) -> bool {
+        self.cmp(right) == Ordering::Equal
+    }
+}
+
+impl<T> Eq for Task<T> {}
+
+impl<T> PartialOrd for Task<T> {
+    fn partial_cmp(&self, rhs: &Task<T>) -> Option<Ordering> {
+        Some(self.cmp(rhs))
+    }
+}
+
+
+pub trait ScheduleQueue<T> {
+    fn pop(&mut self) -> Option<Task<T>>;
+    fn push(&mut self, task: Task<T>);
+    fn finished(&mut self, group_id: T);
+}
+
+// `BigGroupThrottledQueue` tries to throttle group's concurrency to
+//  `group_concurrency_on_busy` when is busy.
+// When one worker asks a task to run, it schedule on the following way:
+// 1. Choose the groups with it's running number less than
+//    `group_concurrency_on_busy`.
+// 2. If more than one groups meet the condition 1, running the one who
+//    comes first.
+// If no group's running number is smaller than `group_concurrency_on_busy`,
+// choose according to the following rules:
+// 1. Choose the groups with least running tasks.
+// 2. If more than one groups has the least running tasks,
+//    choose the one whose task comes first(with the minum task's ID)
+pub struct BigGroupThrottledQueue<T> {
+    // tasks in pending. tasks in `pending_tasks` have higher prority than
+    // tasks in waiting_queue.
+    pending_tasks: BinaryHeap<Task<T>>,
+    // group_id => tasks array. If `statistics[group_id]` is bigger than
+    // `group_concurrency_on_busy`(which means there may more than
+    // `group_concurrency_on_busy` tasks of the group are running), the rest
+    // of the group's tasks would be pushed into `waiting_queue[group_id]`
+    waiting_queue: HashMap<T, VecDeque<Task<T>>>,
+    // group_id => running_num+pending num. It means there may `statics[group_id]`
+    // tasks of the group are running.
+    statistics: HashMap<T, usize>,
```

Comment:
I think statistics should contain more information than the only count here, maybe we should think a better name.